src/bitbox02.js

```diff
@@ -99,87 +113,72 @@ export class BitBox02API {
-                            // Pairing skipped.
-                            break;
-                        default:
-                            throw new Error("Unexpected status: " + self.firmware().Status() + "," + constants.Status.Unpaired);
-                    }
-
-                    resolve();
-                } catch(err) {
-                    reject(err);
-                }
+        this.fw = api.New(onWrite);
+
+        // Turn all Async* methods into promises.
+        for (const key in this.firmware().js) {
+            if (key.startsWith("Async")) {
+                this.firmware().js[key] = promisify(this.firmware().js[key]);
             }
-            self.socket.onerror = function(event) {
-                reject("Your BitBox02 is busy");
+        }
+
+        this.firmware().SetOnEvent(ev => {
+            if (ev === constants.Event.StatusChanged && this.firmware()) {
+                setStatusCb(this.firmware().Status());
             }
-            self.socket.onmessage = function(event) {
-                self.firmware().js.OnRead(new Uint8Array(event.data));
+            if (ev === constants.Event.StatusChanged && this.firmware().Status() === constants.Status.Unpaired) {
+                const [channelHash] = this.firmware().ChannelHash();
+                showPairingCb(channelHash);
             }
-            self.socket.onclose = function(event) {
-                onCloseCb();
+            if (ev === constants.Event.AttestationCheckDone) {
+                handleAttestationCb(this.firmware().Attestation());
+            }
+            if (ev === constants.Event.StatusChanged && this.firmware().Status() === constants.Status.RequireFirmwareUpgrade) {
+                this.socket.close();
+                throw new Error('Firmware upgrade required');
+            }
+            if (ev === constants.Event.StatusChanged && this.firmware().Status() === constants.Status.RequireAppUpgrade) {
+                this.socket.close();
+                throw new Error('Unsupported firmware');
+            }
+            if (ev === constants.Event.StatusChanged && this.firmware().Status() === constants.Status.Uninitialized) {
+                this.socket.close();
+                throw new Error('Uninitialized');
             }
         });
+
+        await this.firmware().js.AsyncInit();
+        switch(this.firmware().Status()) {
+            case constants.Status.PairingFailed:
+                this.socket.close();
+                throw new Error("Pairing rejected");
+            case constants.Status.Unpaired:
+                await userVerify()
+                await this.firmware().js.AsyncChannelHashVerify(true);
+                break;
+            case constants.Status.Initialized:
+                // Pairing skipped.
+                break;
+            default:
+                throw new Error("Unexpected status: " + this.firmware().Status() + "," + constants.Status.Unpaired);
```

Comment:
Single quote or backtick:
`Unexpected status: ${this.firmware().Status()},${constants.Status.Unpaired}`

Reply:
i am just moving code around ;) these improvements are for another PR.